src/main/java/com/conveyal/datatools/manager/jobs/MergeFeedsJob.java

```diff
@@ -455,33 +455,71 @@ private int constructMergedTable(Table table, List<FeedToMerge> feedsToMerge,
-                                // Check that the first file contains stop_code values.
-                                if ("".equals(keyValue)) {
+                            if (lineNumber == 0) {
+                                // Before reading any lines in stops.txt, first determine whether all records contain
+                                // properly filled stop_codes. The rules governing this logic are as follows:
+                                // 1. Stops with location_type greater than 0 (i.e., anything but 0 or empty) are permitted
+                                //    to have empty stop_codes (even if there are other stops in the feed that have
+                                //    stop_code values). This is because these location_types represent special entries
+                                //    that are either stations, entrances/exits, or generic nodes (e.g., for
+                                //    pathways.txt).
+                                // 2. For regular stops (location_type = 0 or empty), all or none of the stops must
+                                //    contain stop_codes. Otherwise, the merge feeds job will be failed.
+                                int stopsMissingStopCodeCount = 0;
+                                int stopsCount = 0;
+                                int locationTypeIndex = getFieldIndex(fieldsFoundInZip, "location_type");
+                                int stopCodeIndex = getFieldIndex(fieldsFoundInZip, "stop_code");
+                                // Get special stops reader to iterate over every stop and determine if stop_code values
+                                // are present.
+                                CsvReader stopsReader = table.getCsvReader(feed.zipFile, null);
+                                while (stopsReader.readRecord()) {
+                                    stopsCount++;
+                                    String locationType = stopsReader.get(locationTypeIndex);
+                                    // "Regular" stop records (i.e., not a station, entrance, or anything with
+                                    // location_type > 0) should contain stop_code.
+                                    boolean isRegularStop = "".equals(locationType) || "0".equals(locationType);
+                                    String stopCode = stopsReader.get(stopCodeIndex);
+                                    boolean stopCodeIsMissing = "".equals(stopCode);
+                                    if (isRegularStop && stopCodeIsMissing) stopsMissingStopCodeCount++;
+                                }
+                                LOG.info("total stops: {}", stopsCount);
+                                LOG.info("stops missing stop_code: {}", stopsMissingStopCodeCount);
+                                if (stopsMissingStopCodeCount == stopsCount) {
+                                    // If all stops are missing stop_code, we simply default to merging on stop_id.
```

Comment:
This doesn't appear to default to using the stop_id anymore, but rather fails the job.

Reply:
That's not exactly true, but I inserted some comments that hopefully describe the failure conditions a little better.